### PR TITLE
Fix user role synchronization and prepare maestro for deployment.

### PR DESCRIPTION
**Pull request by alexperovich**
This change adds the Service Fabric cluster template and supporting files to deploy Maestro++.
It also includes the final changes to enable Maestro++ to function outside of local machine scenarios.

### CODE REVIEW

File: NuGet.Config [hunk 1]
```diff
@@ -7,5 +7,6 @@
   <packageSources>
     <clear />
     <add key="roslyn-tools" value="https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json" />
```
Comment (jcagme) [inline NuGet.Config]: We should change this to something more dotnetfeedy or something in the lines

File: src/Microsoft.DotNet.ServiceFabric.ServiceHost/ServiceHost.cs [hunk 2]
```diff
@@ -250,8 +250,27 @@ private static string GetApplicationName()
 
     public static class ServiceHostConfiguration
     {
-        public static KeyVaultClient GetKeyVaultClient(string connectionString)
+        private static string GetAzureServiceTokenProviderConnectionString(IHostingEnvironment env)
         {
+            if (env.IsDevelopment() && RunningInServiceFabric())
+            {
+                var tenantId = "72f988bf-86f1-41af-91ab-2d7cd011db47";
+                var appId = "388be541-91ed-4771-8473-5791e071ed14";
```
Comment (MattGal) [inline src/Microsoft.DotNet.ServiceFabric.ServiceHost/ServiceHost.cs]: does this principal id never actually change?

File: src/Maestro/Maestro.Web/Controllers/WebHooksController.cs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        public BuildAssetRegistryContext Context => _context.Value;
+
+        public WebHooksController(ILogger<WebHooksController> logger, IGitHubTokenProvider gitHubTokenProvider, Lazy<BuildAssetRegistryContext> context)
+        {
+            _context = context;
+            Logger = logger;
+            GitHubTokenProvider = gitHubTokenProvider;
+        }
+
+        public class InstallationEvent
```
Comment (jcagme) [inline src/Maestro/Maestro.Web/Controllers/WebHooksController.cs]: What does installation mean in this context?

File: src/Maestro/Maestro.Web/Startup.Auth.cs [hunk 4]
```diff
@@ -48,44 +45,70 @@ private void ConfigureAuthServices(IServiceCollection services)
                     {
                         IConfigurationSection ghAuthConfig = Configuration.GetSection("GitHubAuthentication");
                         ghAuthConfig.Bind(options);
-                        options.Events = new OAuthEvents {OnCreatingTicket = AddOrganizationRoles};
+                        options.Events = new OAuthEvents
+                        {
+                            OnCreatingTicket = async context =>
+                            {
+                                var logger = context.HttpContext.RequestServices.GetRequiredService<ILogger<Startup>>();
+                                logger.LogInformation("Reading user roles from github.");
```
Comment (jcagme) [inline src/Maestro/Maestro.Web/Startup.Auth.cs]: Nit: "github" > "GitHub"

File: src/Maestro/Maestro.Web/Startup.Auth.cs [hunk 5]
```diff
@@ -148,30 +173,89 @@ private void ConfigureAuthServices(IServiceCollection services)
...
+                        {
+                            policy.RequireAuthenticatedUser()
+                                .RequireRole("github:team:dotnet:dnceng");
+                        });
+                });
+
+            services.Configure<MvcOptions>(
+                options =>
+                {
+                    options.Conventions.Add(new DefaultAuthorizeActionModelConvention("msft"));
```
Comment (jcagme) [inline src/Maestro/Maestro.Web/Startup.Auth.cs]: Define this in a constant and use the same in line 195 and here

File: src/Maestro/cluster-template/deploy.ps1 [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    Deploys a service fabric cluster
+#>
+
+param(
+ [Parameter(Mandatory=$True)]
+ [string]
+ $subscriptionName,
+
+ [string]
+ $location,
```
Comment (jcagme) [inline src/Maestro/cluster-template/deploy.ps1]: This is not mandatory but I don't think we are checking for its value to be set. Do we need that check?

File: NuGet.Config (hunk 1 above)
Comment (mmitche) [reply NuGet.Config]: I think the key name should be dotnet-core at least.

File: src/Maestro/Maestro.Web/Controllers/WebHooksController.cs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            public Installation Installation { get; set; }
+            public List<InstallationRepository> Repositories { get; set; }
+            public User Sender { get; set; }
+        }
+
+        public class InstallationRepository
+        {
+            public int Id { get; set; }
+            public string Name { get; set; }
+            public string FullName { get; set; }
```
Comment (mmitche) [inline src/Maestro/Maestro.Web/Controllers/WebHooksController.cs]: Is this the URI?

File: src/Maestro/Maestro.Web/Controllers/WebHooksController.cs [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Octokit;
+using Octokit.Internal;
+
+namespace Maestro.Web.Controllers
+{
+    public class WebHooksController : Controller
```
Comment (mmitche) [inline src/Maestro/Maestro.Web/Controllers/WebHooksController.cs]: What's the workflow use for this controller?

File: src/Maestro/Maestro.Web/Startup.Auth.cs [hunk 9]
```diff
@@ -148,30 +173,89 @@ private void ConfigureAuthServices(IServiceCollection services)
...
                     };
                 });
+
+            services.AddAuthorization(
+                options =>
+                {
+                    options.AddPolicy("msft", policy =>
+                        {
+                            policy.RequireAuthenticatedUser()
+                                .RequireRole("github:team:dotnet:dnceng");
```
Comment (mmitche) [inline src/Maestro/Maestro.Web/Startup.Auth.cs]: What does this tie to dnceng?

File: src/Maestro/cluster-template/deploy.ps1 (hunk 6 above)
Comment (alexperovich) [reply src/Maestro/cluster-template/deploy.ps1]: Yea, I will add mandatory to this parameter.

File: src/Maestro/Maestro.Web/Controllers/WebHooksController.cs (hunk 7 above)
Comment (alexperovich) [reply src/Maestro/Maestro.Web/Controllers/WebHooksController.cs]: This is `dotnet/arcade` where Name is `arcade`

File: src/Maestro/Maestro.Web/Controllers/WebHooksController.cs (hunk 8 above)
Comment (alexperovich) [reply src/Maestro/Maestro.Web/Controllers/WebHooksController.cs]: This controller is responsible for responding to WebHook calls from github. Right now it only responds to installation change events to maintain the list of repositories and associated installation ids.

File: src/Maestro/Maestro.Web/Startup.Auth.cs (hunk 9 above)
Comment (alexperovich) [reply src/Maestro/Maestro.Web/Startup.Auth.cs]: This restricts all apis to members of @dotnet/dnceng right now. Once I am able to get the GitHub app permissions in the Microsoft org then I will change this to use that.

File: NuGet.Config [hunk 10]
```diff
@@ -1,11 +1,12 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
 <configuration>
   <solution>
     <add key="disableSourceControlIntegration" value="true" />
   </solution>
   <!-- Only specify feed for Arcade SDK (see https://github.com/Microsoft/msbuild/issues/2982) -->
   <packageSources>
     <clear />
-    <add key="roslyn-tools" value="https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json" />
```
Comment (weshaggard) [inline NuGet.Config]: @alexperovich why did we need to add nuget.org to this? If we need it for more than the arcade sdk then we should update the comment. 

cc @tmat

**Pull request merged**